src/include/mpir_refcount_pobj.h

```diff
@@ -10,45 +10,24 @@
 /* define a type for the completion counter */
 #include "opa_primitives.h"
 
-typedef OPA_int_t MPIR_cc_t;
+typedef MPL_atomic_int_t MPIR_cc_t;
 
-/* implies no barrier, since this routine should only be used for request
- * initialization */
 static inline void MPIR_cc_set(MPIR_cc_t * cc_ptr, int val)
 {
-    if (val == 0) {
-        /* values other than 0 do not enforce any ordering, and therefore do not
-         * start a HB arc */
-        /* MT FIXME using cc_set in this way is sloppy.  Sometimes the caller
-         * really does know that the cc value may cleared, but more likely this
-         * is just a hack to avoid the work of figuring out what the cc value
-         * currently is and decrementing it instead. */
-        /* barrier ensures that any state written before indicating completion is
-         * seen by the thread polling on the cc.  If OPA adds store-release
-         * semantics, we can convert to that instead. */
-        OPA_write_barrier();
-        MPL_VG_ANNOTATE_HAPPENS_BEFORE(cc_ptr);
-    }
-#if defined(MPL_VG_AVAILABLE)
-    /* MT subtle: store_int is actually safe to use, but Helgrind/DRD/TSan all
-     * view the store/load pair as a race.  Using an atomic operation for the
-     * store side makes all three happy.  DRD & TSan also support
-     * ANNOTATE_BENIGN_RACE, but Helgrind does not. */
-    OPA_swap_int(cc_ptr, val);
-#else
-    OPA_store_int(cc_ptr, val);
-#endif
+    /* Using relaxed semantics since this routine should only be used for
+     * request initialization. */
+    MPL_atomic_relaxed_store_int(cc_ptr, val);
 }
 
 ATTRIBUTE((unused))
 static inline int MPIR_cc_is_complete(MPIR_cc_t * cc_ptr)
 {
     int complete;
 
-    complete = (0 == OPA_load_int(cc_ptr));
+    complete = (0 == MPL_atomic_load_int(cc_ptr));
     if (complete) {
-        MPL_VG_ANNOTATE_HAPPENS_AFTER(cc_ptr);
-        OPA_read_barrier();
+        /* REVIEW: is this necessary? */
+        MPL_atomic_read_barrier();
```

Comment:
This read barrier should not be necessary.  The atomic load_int above should already provide both a load and a store barrier (https://en.cppreference.com/w/cpp/atomic/memory_order).

Reply:
removed.